Add ability to output all keys and values in thread context (v3.0 branch)

rhanneken opened the pull request:
### Description

The keys in a thread context can differ for different code paths. It would be good if one could log whatever keys and values exist in the thread context without specifying all possible keys in the logging configuration.

This change modifies the "context" placeholder so that it can be used without specifying a key. When no key is specified, all keys and their values are sent to the output.

Linked website pull request, if the documention of https://tinylog.org/v3/ has to be updated:

tinylog-org/website#196 <!-- Related pull request from https://github.com/tinylog-org/website/pulls -->

### Definition of Done

- [x] I read [contributing.md](https://github.com/tinylog-org/tinylog/blob/v2.8/contributing.md)
- [x] There are no TODOs left in the code
- [x] Code style follows the tinylog standard
- [x] All classes and methods have Javadoc
- [x] Changes are covered by JUnit tests including edge cases, errors, and exception handling
- [x] Maven build works including compiling, tests, and checks (`mvn verify`)
- [x] Changes are committed by a verified email address that is assigned to the GitHub account (https://github.com/settings/emails)
- [x] Documentation on the [tinylog website](https://tinylog.org/v2/) is up-to-date or updated by a pull request to the repository [tinylog-org/website](https://github.com/tinylog-org/website)

### Agreements

- [x] I agree that my changes will be published under the terms of the [Apache License 2.0](https://github.com/tinylog-org/tinylog/blob/v2.0/license.txt) (mandatory)
- [x] I agree that my GitHub user name will be published in the release notes (optional)

rhanneken committed:
Add ability to output all keys and values in thread context

The keys in a thread context can differ for different code paths. It
would be good if one could log whatever keys and values exist in the
thread context without specifying all possible keys in the logging
configuration.

This change modifies the "context" placeholder so that it can be used
without specifying a key. When no key is specified, all keys and their
values are sent to the output.

pmwmedia commented on tinylog-impl/src/main/java/org/tinylog/impl/format/pattern/placeholders/ContextPlaceholder.java:
```diff
@@ -33,13 +42,36 @@ public ValueType getType() {
+
+        return contextEntries.stream()
+            .sorted(Map.Entry.comparingByKey())
+            .map(me -> new StringBuilder().append(me.getKey()).append("=").append(me.getValue()))
```
Could you rename `me` into `mapEntry` and `entry` into `logEntry`?

rhanneken replied:
Done.

pmwmedia commented on tinylog-impl/src/main/java/org/tinylog/impl/format/pattern/placeholders/ContextPlaceholder.java:
```diff
@@ -33,13 +42,36 @@ public ValueType getType() {
+     * @param defaultValue the value to return if the thread context has no keys or values
+     * @return a string representation of all keys and values in entry's thread context
+     */
+    private static String contextKeysAndValuesOrDefault(LogEntry entry, String defaultValue) {
```
Could you start Javadoc descriptions with an upper letter? This makes it easy to distinguish the actual param from its description.

```
* @param entry The log entry whose thread context is to be represented as a string
* @param defaultValue The value to return if the thread context has no keys or values
* @return A string representation of all keys and values in entry's thread context
```

rhanneken replied:
Done.

pmwmedia commented on tinylog-impl/src/main/java/org/tinylog/impl/format/pattern/placeholders/ContextPlaceholder.java:
```diff
@@ -33,13 +42,36 @@ public ValueType getType() {
+    private static String contextKeysAndValuesOrDefault(LogEntry entry, String defaultValue) {
+        Set<Map.Entry<String, String>> contextEntries = entry.getContext().entrySet();
+        if (contextEntries.isEmpty()) {
+            return defaultValue;
```
I think we can remove `defaulValue` as method parameter and return always an empty string `""` here.

rhanneken replied:
Done.

pmwmedia reviewed (commented):
Thank you for your pull request! I really like your migration for tinylog 3. I have only commented a few minor improvements, which should be quick and easy to apply.

One general idea: In tinylog 3, we use builders for the creation of placeholders, which would allow the creation of different placeholders by the same builder. What do you think about splitting the logic into two different placeholders? One placeholder for the old logic (outputting the value for a key) and another new placeholder for outputting all thread context entries.

rhanneken committed:
Address code review feedback

A previous commit introduced an ehancement allowing the user to output
all thread context keys and values by using the "context" placeholder
without specifying a key. This commit addresses feedback from a review
of that code. Notably, the ContextPlaceholder class has been refactored
into a class hierarchy: an abstract base class, a subclass that
encapsulates the logic of a context placeholder with a key, and a
subclass that encapsulates the logic of a context placeholder without a
key.

rhanneken committed:
Renamed variable as suggested by code review

rhanneken commented:
>What do you think about splitting the logic into two different placeholders? One placeholder for the old logic (outputting the value for a key) and another new placeholder for outputting all thread context entries.

That's a good idea, and I've implemented your suggestion. Let me know what you think.

pmwmedia commented on tinylog-impl/src/main/java/org/tinylog/impl/format/pattern/placeholders/ContextPlaceholderBuilder.java:
```diff
@@ -18,11 +18,6 @@ public String getName() {
-        } else {
-            return new ContextPlaceholder(value);
-        }
+        return value != null ? new SingleValueContextPlaceholder(value) : new MultiValueContextPlaceholder();
```
I would prefer the way around:

```java
return value == null ? new MultiValueContextPlaceholder() : new SingleValueContextPlaceholder(value);
```

rhanneken replied:
OK, changed.

pmwmedia commented on tinylog-impl/src/main/java/org/tinylog/impl/format/pattern/placeholders/MultiValueContextPlaceholder.java:
```diff
@@ -0,0 +1,49 @@
+
+        return contextEntries.stream()
+            .sorted(Map.Entry.comparingByKey())
+            .map(mapEntry -> new StringBuilder().append(mapEntry.getKey()).append("=").append(mapEntry.getValue()))
```
Can be reduced to:

```java
new StringBuilder(mapEntry.getKey()).append('=').append(mapEntry.getValue())
```

rhanneken replied:
Good point. Changed.

pmwmedia reviewed (commented):
Thank you for the refactoring! The code looks very clean :) I found just two last minors.

rhanneken committed:
Minor changes due to code review feedback

rhanneken commented:
> I found just two last minors.

I've addressed them in my latest commit. Let me know what you think.

pmwmedia commented:
Thank you for your quick commit. I'm happy to merge the PR :)
